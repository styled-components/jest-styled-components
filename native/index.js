const toHaveStyleRule = require('../src/matchers/toHaveStyleRule')

expect.extend({ toHaveStyleRule })
