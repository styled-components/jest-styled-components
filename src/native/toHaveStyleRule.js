const { matcherTest, buildReturnMessage } = require('../utils');

function toHaveStyleRule({ props: {style} }, property, expected) {
  const styles = Array.isArray(style) ? style.filter(x => x) : style

  /**
   * Convert style name to camel case (so we can compare)
   */
  const camelCasedProperty = property.replace(/-(\w)/, (_, match) => match.toUpperCase());

  /**
   * Merge all styles into one final style object and search for the desired
   * stylename against this object
   */
  const mergedStyles =
    Array.isArray(styles)
      ? styles.reduce((acc, item) => (Object.assign({}, acc, item)), {})
      : styles
  const received = mergedStyles[camelCasedProperty];
  const pass = matcherTest(received, expected, this.isNot);

  return {
    pass,
    message: buildReturnMessage(this.utils, pass, property, received, expected),
  };
}

module.exports = toHaveStyleRule;
