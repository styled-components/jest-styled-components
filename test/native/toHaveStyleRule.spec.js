import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import styled, { ThemeProvider } from 'styled-components/native'
import '../../native'

test('basic', () => {
  const StyledView = styled.View`
    background-color: papayawhip;
  `
  const tree = renderer.create(<StyledView />).toJSON()

  expect(tree).toHaveStyleRule('background-color', 'papayawhip')
})

test('message when property not found', () => {
  const Button = styled.Text`
    background-color: papayawhip;
  `

  expect(() =>
    expect(renderer.create(<Button />).toJSON()).toHaveStyleRule(
      'color',
      'black'
    )
  ).toThrowErrorMatchingSnapshot()
})

test('message when value does not match', () => {
  const StyledView = styled.View`
    background-color: orange;
  `

  expect(() => {
    expect(renderer.create(<StyledView />).toJSON()).toHaveStyleRule(
      'background-color',
      'red'
    )
  }).toThrowErrorMatchingSnapshot()
})

test('basic', () => {
  const StyledView = styled.View`
    padding: 4px;
    background-color: papayawhip;
  `

  expect(renderer.create(<StyledView />).toJSON()).toHaveStyleRule(
    'background-color',
    'papayawhip'
  )
})

test('regex', () => {
  const StyledView = styled.View`
    padding: 4px;
    background-color: papayawhip;
  `

  expect(renderer.create(<StyledView />).toJSON()).toHaveStyleRule(
    'background-color',
    /^papaya/
  )
})

test('undefined', () => {
  const Button = styled.Text`
    ${({ transparent }) => !transparent && 'background-color: papayawhip;'};
  `

  expect(renderer.create(<Button />).toJSON()).toHaveStyleRule(
    'background-color',
    'papayawhip'
  )
  expect(renderer.create(<Button transparent />).toJSON()).toHaveStyleRule(
    'background-color',
    undefined
  )
})

test('jest asymmetric matchers', () => {
  const Button = styled.Text`
    background-color: ${({ transparent }) =>
      transparent ? 'transparent' : 'papayawhip'};
  `

  expect(renderer.create(<Button />).toJSON()).toHaveStyleRule(
    'background-color',
    expect.any(String)
  )
  expect(renderer.create(<Button />).toJSON()).toHaveStyleRule(
    'background-color',
    expect.stringMatching('papayawhip')
  )
  expect(renderer.create(<Button />).toJSON()).toHaveStyleRule(
    'background-color',
    expect.stringMatching(/^papaya/)
  )
  expect(renderer.create(<Button transparent />).toJSON()).toHaveStyleRule(
    'background-color',
    expect.stringContaining('transparent')
  )
  expect(renderer.create(<Button transparent />).toJSON()).not.toHaveStyleRule(
    'color',
    expect.any(String)
  )
  expect(renderer.create(<Button transparent />).toJSON()).not.toHaveStyleRule(
    'color',
    expect.anything()
  )
})

test('styled child', () => {
  const Parent = styled.Text`
    color: red;
  `
  const StyledChild = styled(Parent)`
    padding: 0;
  `

  expect(renderer.create(<StyledChild />).toJSON()).toHaveStyleRule(
    'color',
    'red'
  )
})

test('extending styles', () => {
  const Button = styled.Text`
    color: palevioletred;
    font-size: 1px;
    margin: 1px;
    padding: 0.25px 1px;
    border: 2px solid palevioletred;
    border-radius: 3px;
  `

  const TomatoButton = Button.extend`
    color: tomato;
    border-color: tomato;
  `

  expect(renderer.create(<TomatoButton />).toJSON()).toHaveStyleRule(
    'color',
    'tomato'
  )
})

test('theming', () => {
  const Button = styled.Text`
    font-size: 1px;
    margin: 1px;
    padding: 0.25px 1px;
    border-radius: 3px;

    color: ${props => props.theme.main};
    border: 2px solid ${props => props.theme.main};
  `

  Button.defaultProps = {
    theme: {
      main: 'palevioletred',
    },
  }

  const theme = {
    main: 'mediumseagreen',
  }

  expect(renderer.create(<Button>Normal</Button>).toJSON()).toHaveStyleRule(
    'color',
    'palevioletred'
  )

  const component = (
    <ThemeProvider theme={theme}>
      <Button>Themed</Button>
    </ThemeProvider>
  )

  expect(renderer.create(component).toJSON()).toHaveStyleRule(
    'color',
    'mediumseagreen'
  )
})
