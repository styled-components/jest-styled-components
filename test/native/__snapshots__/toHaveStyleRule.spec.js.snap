// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`message when property not found 1`] = `
"[31m\\"Property 'color' not found in style rules\\"[39m

Expected
  [32m\\"color: black\\"[39m
Received:
  [31m\\"color: undefined\\"[39m"
`;

exports[`message when value does not match 1`] = `
"[31m\\"Value mismatch for property 'background-color'\\"[39m

Expected
  [32m\\"background-color: red\\"[39m
Received:
  [31m\\"background-color: orange\\"[39m"
`;
